Feature/huawei support

Abdullqadir opened the pull request:
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Added Huawei support as mentioned in this issue #81.

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected -->

### Why Should This Be In Core?

To support those devices that do not have google play service.

### Benefits

Huawei Support

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed, commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues

<!-- Enter any applicable Issues here -->

Abdullqadir committed:
Add Huawei mobile service location dependencies to the project.

Abdullqadir committed:
Fix issue 81, Support Huawei Mobile Service Location Kit.

Abdullqadir commented:
I have faced some difficulties while trying to register the sample project in the Huawei App Gallery saying the package name is already registered.

Abdullqadir commented:
Any Updates?

BirjuVachhani commented:
Hey @Abdullqadir 
Thank you for opening this PR! I'll review it as soon as I have some bandwidth!

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/LocusActivity.kt:
```diff
@@ -66,10 +74,10 @@ class LocusActivity : AppCompatActivity(), ActivityCompat.OnRequestPermissionsRe
 
         intent?.getParcelableExtra<Configuration>(Constants.INTENT_EXTRA_CONFIGURATION)?.let {
-            config = it
+            config = it.copy(locationRequest = getLocusLocationRequest(getAvailableService()))
```
Why are you calling `copy` and resetting `LocationRequest` to default configuration?

Abdullqadir replied:
Done.

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/Locus.kt:
```diff
@@ -125,6 +128,7 @@ object Locus {
      */
     @Throws(IllegalStateException::class)
     fun startLocationUpdates(context: Context): LiveData<LocusResult> {
+        config.locationRequest = getLocusLocationRequest(context.getAvailableService())
```
Am I understanding this wrong or this just overrides locationRequest with default settings?

Abdullqadir replied:
It will update the value of the locationRequest with the available service in this context.

BirjuVachhani replied:
I get that. However, what happens to user's custom request? This uses `getHMSLocationRequest` and `getGMSLocationRequest` methods which always returns a new instance of the request. User's custom settings are gone. You're not supposed to do that!

<img width="641" alt="image" src="https://github.com/BirjuVachhani/locus-android/assets/20423471/b00c4809-e684-4565-9c40-afe93b1fdd32">

BirjuVachhani replied:
These are the logs I am getting. It is clear that your code here is resetting user's location request configuration.

Please make sure this never happens!

<img width="726" alt="image" src="https://github.com/BirjuVachhani/locus-android/assets/20423471/9770b7f1-1088-425b-a2c2-21cef5d2d8a2">
<img width="996" alt="image" src="https://github.com/BirjuVachhani/locus-android/assets/20423471/5c092171-9118-4cb7-b4dc-bd829b45b3f7">
<img width="974" alt="image" src="https://github.com/BirjuVachhani/locus-android/assets/20423471/a02ab6b4-a427-4c26-ad61-3e41da60ebd0">

Abdullqadir replied:
Thanks for the details.

Abdullqadir replied:
Hello @BirjuVachhani,
I hope you are doing well.
I have made some changes regarding the user configurations on the Location Request please check it [c7a06fb](https://github.com/BirjuVachhani/locus-android/pull/87/commits/c7a06fb5c757bc4253b4e3f73629d768b35ff41e).

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/Locus.kt:
```diff
@@ -342,8 +386,9 @@ object Locus {
-        onResult: (LocusResult) -> Unit
+        onResult: (LocusResult) -> Unit,
     ) {
+        config.locationRequest = getLocusLocationRequest(context.getAvailableService())
```
Same goes for this too!

Abdullqadir replied:
Updated.

BirjuVachhani commented:
> I have faced some difficulties while trying to register the sample project in the Huawei App Gallery saying the package name is already registered.

@Abdullqadir Is this necessary to run the example app on Huawei devices?

Abdullqadir commented:
> > I have faced some difficulties while trying to register the sample project in the Huawei App Gallery saying the package name is already registered.
> 
> @Abdullqadir Is this necessary to run the example app on Huawei devices?

Yes, it is.

BirjuVachhani commented:
> > > I have faced some difficulties while trying to register the sample project in the Huawei App Gallery saying the package name is already registered.
> > 
> > 
> > @Abdullqadir Is this necessary to run the example app on Huawei devices?
> 
> Yes, it is.

Interesting! I just tested a debug build on a real Huawei device using [BrowserStack](https://www.browserstack.com/) and it works with your current config on the PR! Idk if removing them would make it not work or something!

To me, it seems that you should be able to run any debug builds on Huawei devices w/o doing this config. Seems like it is only required if you wanna publish to their app store or something!

Abdullqadir commented:
> > > > I have faced some difficulties while trying to register the sample project in the Huawei App Gallery saying the package name is already registered.
> > > 
> > > 
> > > @Abdullqadir Is this necessary to run the example app on Huawei devices?
> > 
> > 
> > Yes, it is.
> 
> Interesting! I just tested a debug build on a real Huawei device using [BrowserStack](https://www.browserstack.com/) and it works with your current config on the PR! Idk if removing them would make it not work or something!
> 
> To me, it seems that you should be able to run any debug builds on Huawei devices w/o doing this config. Seems like it is only required if you wanna publish to their app store or something!

That is great. If it works then we shall proceed.
I will resolve your comments to finalize everything.
Regarding the testing, I have only tested the sample application. I will test it again tomorrow.

Abdullqadir committed:
Remove the copy method call on the Configuration data class after getting it back from the intent as it is unnecessary.

Abdullqadir committed:
Bug Fix, After installing the application for the first time the fused location provider returns null on success listener.

Abdullqadir committed:
Fix issue 81, Support Huawei Mobile Service Location Kit, Remove the hms client appid and cpid from the manifest as we don't need it.

Abdullqadir commented:
Hello @BirjuVachhani,
I have made some updates and answered your questions please check them.
If you have any concerns or anything you want to change please let me know.

Abdullqadir commented:
Regarding testing, I have tested everything it works fine. I have only faced one issue which was:

> After installing the application for the first time if we press the Single Update button before the Start update button we were getting back null on the success listener on the fused location provider.

I have fixed this issue in this commit [ab71367](https://github.com/BirjuVachhani/locus-android/pull/87/commits/ab71367d29381d7a8cea025e08183715e0b31ac7).

Abdullqadir committed:
Fix issue 81, Support Huawei Mobile Service Location Kit, Fix ignoring user's configurations on location request bug.

Abdullqadir commented:
Hello @BirjuVachhani,
I hope that you are doing well.
I have fixed the issue regarding the user's configurations on Location Request I have updated it and tested it on Huawei and Google. works fine.
Please let me know if you have any concerns.

BirjuVachhani commented:
@Abdullqadir Thanks for the update. I'll review it.

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/LocusActivity.kt:
```diff
@@ -66,10 +74,10 @@ class LocusActivity : AppCompatActivity(), ActivityCompat.OnRequestPermissionsRe
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_location_permission)
-
+        config = Configuration(getLocusLocationRequest(getAvailableService()))
```
No need to initialize config here as it should parsed from the Intent!

Abdullqadir replied:
I have removed this line already.

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/Configuration.kt:
```diff
@@ -44,23 +50,123 @@ data class Configuration(
-}
+@Parcelize
+data class LocationRequestConfiguration(
+    var priority: Int = Priority.PRIORITY_BALANCED_POWER_ACCURACY,
```
The default value for priority was `PRIORITY_HIGH_ACCURACY` before. Please don't change default values. It may affect a lot of apps.

Abdullqadir replied:
Updated

BirjuVachhani commented on locus/src/main/java/com/birjuvachhani/locus/Configuration.kt:
```diff
@@ -44,23 +50,123 @@ data class Configuration(
+    var fastestInterval: Long = FASTEST_INTERVAL_IN_MS,
+    var expirationTime: Long = Long.MAX_VALUE,
+    var numUpdates: Int = Int.MAX_VALUE,
+    var smallestDisplacement: Float = 0.0F,
```
Make sure this is the default value used by LocationRequest object!

Abdullqadir replied:
In GMS and HMS documentation the default value is zero.

BirjuVachhani commented:
@Abdullqadir  Nice job on the PR, specially on `LocationRequestConfiguration`.

Abdullqadir commented:
> @Abdullqadir Nice job on the PR, specially on `LocationRequestConfiguration`.

Thanks, @BirjuVachhani, I will resolve everything ASAP.

Abdullqadir committed:
FiFix issue 81, Support Huawei Mobile Service Location Kit.

Abdullqadir commented:
Hello @BirjuVachhani, I have resolved all the comments, Please let me know if there is anything else to update.

BirjuVachhani reviewed (approved):
LGTM

BirjuVachhani commented:
Thank you @Abdullqadir for your contributions. 🎉

I'll try to release this today!

Abdullqadir commented:
> I'll try to release this toda

I'm glad to hear that.
Thanks to you for your reviews @BirjuVachhani
